components/calendar/src/calendar_arithmetic.rs

```diff
@@ -169,13 +174,135 @@ impl<C: DateFieldsResolver> ArithmeticDate<C> {
         ArithmeticDate { year, month, day }
     }
 
-    /// Construct a new arithmetic date from a year, month ordinal, and day, bounds checking
-    /// the month and day according to the overflow parameter.
-    pub(crate) fn try_from_builder(
-        builder: ArithmeticDateBuilder<C::YearInfo>,
+    pub(crate) const fn cast<C2: DateFieldsResolver<YearInfo = C::YearInfo>>(
```

Comment:
Nit: This doesn't seem like a safe cast because we use `i32` as the YearInfo in most calendars, but the i32 has different epochs.

Reply:
the cast here is for things like `Ethiopian`: it parses into a YMD tuple using a Coptic year, but the type is `ArithmeticDate<Ethiopian>`, so we need to cast to store that as a `ArithmeticDate<Coptic>`. 

we could avoid this is we made `ArithmeticDate` *not* be generic in a `CalendarArithmetic`, but instead generic in a `YearInfo`. Then the constructors would each be generic, and we'd store `ArithmeticDate<i32>` or `ArithmeticDate<HijriYearData>`.